compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -1558,181 +1083,696 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             );
         }
 
-        if rcvr_ty.is_numeric() && rcvr_ty.is_fresh() || restrict_type_params || suggested_derive {
+        self.note_derefed_ty_has_method(&mut err, source, rcvr_ty, item_ident, expected);
+        err.emit()
+    }
+
+    fn set_not_found_span_label(
+        &self,
+        err: &mut Diag<'_>,
+        rcvr_ty: Ty<'tcx>,
+        item_ident: Ident,
+        item_kind: &str,
+        mode: Mode,
+        source: SelfSource<'tcx>,
+        span: Span,
+        unsatisfied_predicates: &Vec<(
+            ty::Predicate<'tcx>,
+            Option<ty::Predicate<'tcx>>,
+            Option<ObligationCause<'tcx>>,
+        )>,
```

Comment:
This type is repeated multiple times. Can you introduce a new `type` to give it a name? It can be used in a few places outside this file as well, grep for `"unsatisfied_predicates: "`.